### ISSUE TITLE
How horovod loads data into memory

### ISSUE THREAD

**Issue by sudheerExperiments** [horovod/horovod#733]
Hello,

I have a question regarding *'how horovod loads data into memory'*. We are using a cluster with 2 compute nodes, each compute node has 4 GPUs, so in total we have 8 GPUs. One compute node has 16 GB memory and the other has 64 GB memory. We are using Keras to train a CNN model using 'MPICH 3'. Our dataset is around 8.5 GB. We are running 8 processes (1 process per each GPU, 4 processes per one compute node). We were frequently getting 'memory collection' error, so I started keeping track of memory. I noticed something odd.

**Before loading data:**

**Compute node 1:**
RAM usage: 3.11 GB / 15.3 GB
Compute node 2:
RAM usage: 8.09 GB / 63.4 GB

**After loading data:**

**Compute node 1:**
RAM usage: 15.1 GB / 15.3 GB
Compute node 2:
RAM usage: 42.5 GB / 63.4 GB

Since our dataset is 8.5 GB, shouldn't be the memory increase approx. 9 GB or so? Looks like horovod is loading dataset 4 times into memory (for each process). 

How to tackle this issue? If anybody faced similar issue, need help.

Thanks and Regards,
Sudheer.

**Comment by alsrgv** [comment]
@sudheerExperiments, can you refactor your code to load only portion of the dataset on each node?  You can use `hvd.rank()` and `hvd.size()` to carve out the chunk to load on the particular process.

**Comment by sudheerExperiments** [comment]
Sure I will try it out. Thanks @alsrgv.

One question though, even I modify my load part of code into memory, does horovod load data 4 times if I am running 4 processes?

But doesn't *'mpirun'* take care of it?

**Comment by alsrgv** [comment]
@sudheerExperiments, neither mpirun nor Horovod do anything about data loading.  This responsibility lies with the model developer.  If you want to keep the data distributed across memory of all processes, you could have each process load a slice based on `hvd.rank()` and `hvd.size()`.

**Comment by sudheerExperiments** [comment]
Hello @alsrgv,

I tried what you suggested. I used **hvd.rank** to split my dataset. I have training data of shape ```(40, 90, 90)``` and labels of shape ```(40, 1)```. I seperated it as 8 ```(5, 90, 90)``` and ```(5, 1)``` sets. When **'mpirun'** launches 8 GPU jobs using horovod, it gets seperate data for each process. How does horovod uses these datasets to train? Does it treat each set as mini-batch, calculates gradiants and updates weights one by one or does it do something else? Its not clear from my training results, please help.

Thanks in advance.

**Comment by abditag2** [comment]
@sudheerExperiments In each step of training, each rank creates a batch from its slices of data, calculates local gradients and then horovod calculates the global gradients and distributes it back to the ranks. The weights on all the ranks are updated together in synchronous steps.

**Comment by sudheerExperiments** [comment]
Hello @abditag2, Thanks! I researched more about horovod. Now its much clear to me how horovod works. Dividing data into parts is a big challenge to me now. Thanks.

**Comment by sedghi** [comment]
I have a question regarding the same process
I'm using horovod with 2 GPUs on one computer. I followed all the instructions but horovod loads the dataset twice on the memory. Is there any explanation of why this happens? 
I thought it loads the data into the memory ONCE, and send portions of it to each model on each GPU.

**Comment by alsrgv** [comment]
@sedghi, I believe recent Keras supports tf.data API, which does in turn support sharding, so you could use that.  This comment mentions shard API: https://github.com/uber/horovod/issues/708#issuecomment-447664883

**Comment by Leslie-Fang** [comment]
> @sudheerExperiments In each step of training, each rank creates a batch from its slices of data, calculates local gradients and then horovod calculates the global gradients and distributes it back to the ranks. The weights on all the ranks are updated together in synchronous steps.

Hi @abditag2 Thanks for the reply.
I have another question is is it possible for each rank creates a batch of data meanwhile updates the weights. I think these 2 steps are irreverent, Right? After both of this steps are finished, the new step could start the forward pass.

**Comment by Leslie-Fang** [comment]
https://github.com/horovod/horovod/issues/1840
Thanks this reply resolves my problem

**Comment by skrsna** [comment]
Hi @alsrgv , 
I'm seeing this with pytorch's data loaders, is there an example on how to do data sharding with custom DataLoader in pytorch. My datasets are around ~100G and I'm using  Horovod on 4 V100s. The training scripts keeps getting killed once the system memory reaches ~50G during data loading. Thanks for such an awesome tool.

**Comment by tgaddair** [comment]
Hey @skrsna, what format is your data in?  For very large datasets, we recommend putting them into a shardable format like Parquet, then using a framework like [Petastorm](https://github.com/uber/petastorm) to stream it in across workers.

**Comment by skrsna** [comment]
Hey @tgaddair, my datasets are [DGL](https://www.dgl.ai) graphs with features on nodes and edges that are torch tensors. I'll try out Petastorm. Thanks
